tools/codegen/gen.py

```diff
@@ -312,31 +312,65 @@ def go(f: NativeFunction) -> Optional[str]:
             return None
 
         name = cpp.name(f.func)
-
         cpp_returns_type = cpp.returns_type(f.func.returns)
-        cpp_args = cpp.arguments(f.func)
-        cpp_args_str = ', '.join(map(str, cpp_args))
+        cpp_binding = cpp.binding(f.func)
 
         if target is Target.DECLARATION:
-            return f"CAFFE2_API {cpp_returns_type} {name}({cpp_args_str});"
+            if cpp_binding.gathered_signature is None:
+                # There's no TensorOptions
+                result = f"CAFFE2_API {cpp_returns_type} {name}({cpp_binding.signature.cpp_arguments_str()});"
+            else:
+                result = f"CAFFE2_API {cpp_returns_type} {name}({cpp_binding.gathered_signature.cpp_arguments_str()});\n"
+                result += f"CAFFE2_API {cpp_returns_type} {name}({cpp_binding.signature.no_defaults().cpp_arguments_str()});"
+
+            return result
 
         assert target is Target.DEFINITION
 
-        dispatcher_exprs = dispatcher.cpparguments_exprs(cpp_args)
-        cpp_args_str_no_default = ', '.join(map(lambda a: a.str_no_default(), cpp_args))
         dispatcher_returns_type = dispatcher.returns_type(f.func.returns)
-        dispatcher_types_str = ', '.join(map(lambda a: a.type, dispatcher_exprs))
-        dispatcher_exprs_str = ', '.join(map(lambda a: a.expr, dispatcher_exprs))
 
-        return f"""
+        if cpp_binding.gathered_signature is None:
+            # There's no TensorOptions
+            return f"""
 // aten::{f.func}
-{cpp_returns_type} {name}({cpp_args_str_no_default}) {{
+{cpp_returns_type} {name}({cpp_binding.signature.no_defaults().cpp_arguments_str()}) {{
     static auto op = c10::Dispatcher::singleton()
         .findSchemaOrThrow("aten::{f.func.name.name}", "{f.func.name.overload_name}")
-        .typed<{dispatcher_returns_type} ({dispatcher_types_str})>();
-    return op.call({dispatcher_exprs_str});
+        .typed<{dispatcher_returns_type} ({cpp_binding.signature.types_str()})>();
+    return op.call({cpp_binding.signature.exprs_str()});
+}}
+"""
+        elif local.use_c10_dispatcher() is UseC10Dispatcher.full:
+            # for c10-full ops, the scattered version is the real op and the gathered version is a proxy
+            # calling into the scattered version
```

Comment:
I'm not sure why this proxying business is going on here. Wouldn't it be simpler to just define translation from "signature to dispatcher" and from "gathered signature to dispatcher" and call it a day? Then you would have only a *single* calling template to populate (the dispatcher code) and you can dedupe the four variants below. I agree that for the signatures above you still need to special case because of defaulting behavior, but for the inside implementations of these functions, I'm pretty sure the template should be the same in all situations.

Reply:
Let me try to elucidate the underlying design constraint here. I don't think there is anything fundamentally wrong with proxying gathered signature to scattered signature. But it introduces a ton of new translations:

This PR adds:
* Scattered cpp -> gathered cpp translation
* Gathered cpp -> scattered cpp translation
* Gathered cpp -> dispatcher translation
* Scattered cpp -> dispatcher translation

That's a ton of stuff! It seems conceptually simpler to me to reduce the number of translations you need:

* Scattered cpp -> dispatcher translation
* Gathered cpp -> dispatcher translation

You don't need the scattered cpp to gathered cpp translation, because you can just go *directly* to dispatcher. It also makes it harder to make mistakes, as you don't have scattered/gather cpp in the covariant return position, so you don't have to worry about mixing the two of them up.

Reply:
this doesn't work unfortunately. Even if you go directly to the dispatcher, you'll need all 4 translations. There's 2 C++ API calling conventions and 2 dispatcher calling conventions and you need all 4 edges between those nodes.